src/NHibernate/Type/OneToOneType.cs

```diff
@@ -135,25 +154,41 @@ public override bool UseLHSPrimaryKey
 
 		public override object Disassemble(object value, ISessionImplementor session, object owner)
 		{
-			return null;
+			if (value == null)
+			{
+				return null;
+			}
+
+			object id = ForeignKeys.GetEntityIdentifierIfNotUnsaved(GetAssociatedEntityName(), value, session);
+
+			if (id == null)
+			{
+				throw new AssertionFailure("cannot cache a reference to an object with a null id: " + GetAssociatedEntityName());
+			}
+
+			return GetIdentifierType(session).Disassemble(id, session, owner);
 		}
 
 		public override object Assemble(object cached, ISessionImplementor session, object owner)
 		{
-			//this should be a call to resolve(), not resolveIdentifier(), 
-			//'cos it might be a property-ref, and we did not cache the
-			//referenced value
-			return ResolveIdentifier(session.GetContextEntityIdentifier(owner), session, owner);
+			// the owner of the association is not the owner of the id
+			object id = GetIdentifierType(session).Assemble(cached, session, null);
+
+			if (id == null)
+			{
+				return null;
+			}
+
+			return ResolveIdentifier(id, session);
 		}
 
 		/// <summary>
-		/// We don't need to dirty check one-to-one because of how 
-		/// assemble/disassemble is implemented and because a one-to-one 
-		/// association is never dirty
+		/// We always need to dirty check as our identifier is dependent on
+		/// whether or not a persistent entry exists.
 		/// </summary>
 		public override bool IsAlwaysDirtyChecked
 		{
-			get { return false; } //TODO: this is kinda inconsistent with CollectionType
+			get { return true; }
```

Comment:
For many-to-one it's true only for `not-found="ignore"`. Why it's always true here? Shouldn't it be here similarly true only for `constrained="false"` associations (so basically `return IsNullable;`)

Reply:
There are two use cases for a one-to-one with constrained="false" on the parent side and constrained="true" on the child side. From my understanding of `IsAlwaysDirtyChecked` only determines whether or not the relations should be checked to see if it is dirty. When constrained="false" we definitely want to always perform the dirty check. When constrained="true", the id of the child should match the id of the parent and so presumably it can never be dirty. I think returning `IsNullable` here would be acceptable.